Tests/MongoSwiftTests/CrudTests.swift

```diff
@@ -205,7 +205,7 @@ private class CrudTest {
     /// Given the response to a findAndModify command, verify that it matches the expected
     /// results for this `CrudTest`. Meant for use by findAndModify subclasses, i.e. findOneAndX. 
     func verifyFindAndModifyResult(_ result: Document?) {
-        if self.result == nil {
+        if self.result is NSNull {
```

Comment:
is stuff actually stored as null in the test files rather than being missing?

Reply:
Good question, I'm actually not sure (I clearly assumed that they were at the time of making this change). Is there a way to test this?

My guess is we could take one of the JSON files and mess with it by removing an optional field in the JSON, but I don't actually know which of these fields can be wiped. I've tried a few, but all the fields I've tried just removing have failed both on master and this branch and in the same way, suggesting that those fields are probably mandatory for the test.

Reply:
just look at the find and modify files we are parsing and using this on, and see if they have a result field or have JSON null?

Reply:
Right, but this only shows what is going on with the current files. Would we not want to make sure that in the future any new files added would not omit the result field, or if they do, that it would be represented with `null` instead of just be missing outright?

Reply:
"FindOneAndDelete when no documents match" in `findOneAndDelete.json` is an example where `result` is `null` in JSON. In that case, I think the `findOneAndDelete()` method should actually return `nil` instead of NSNull, since NSNull represents a BSON null value within a document. This may be related to my earlier comment about parsing the CRUD test.

If we're parsing the entire CRUD test as a BSON Document (instead of directly from JSON), then I suppose we may have to interpret an NSNull in `outcome["result"]` as a `nil` for this purpose.

Reply:
Yep, in the [`init()` call](https://github.com/mongodb/mongo-swift-driver/pull/145/files#diff-e8428080876ca5bc2e6ecb19e22d6771L163) for the initialization of `self.result`, we parse in the JSON as a `Document`, so we'd use `NSNull` for JSON `null`.

Reply:
Done.

In the case of the `DistinctTest`, I make it so that if the result is non-`nil`, then we proceed as normal. If `self.result` _is_ `nil`, then I expect an empty document instead. This seems to agree with what `distinct()` returns in the Mongo shell when I use it on something like an empty collection, so I imagine someone leaving out `"result"` in the JSON is implying that the result should be empty.

This is what I've done, for quick reference:
```swift
/// A class for executing `distinct` tests
private class DistinctTest: CrudTest {
    override func execute(usingCollection coll: MongoCollection<Document>) throws {
        let filter = self.args["filter"] as? Document
        let fieldName: String = try self.args.get("fieldName")
        let options = DistinctOptions(collation: self.collation)
        // rather than casting to all the possible BSON types, just wrap the arrays in documents to compare them
        let resultDoc: Document = ["result": try coll.distinct(fieldName: fieldName,
                                                               filter: filter ?? [:],
                                                               options: options)]
        let expectedDoc: Document
        if let result = self.result {
            expectedDoc = ["result": result]
        } else {
            expectedDoc = [:]
        }
        expect(resultDoc).to(equal(expectedDoc))
    }
}
```

In particular:
```swift
        let expectedDoc: Document
        if let result = self.result {
            expectedDoc = ["result": result]
        } else {
            expectedDoc = [:]
        }
        expect(resultDoc).to(equal(expectedDoc))
```

Reply:
all sgtm

Reply:
If `outcome.result` field is missing from the CRUD spec test case's JSON, it would be correct to not assert anything about the return value of the CRUD method. Consider the case where `outcome.error` is `true` and the spec test only wants to assert that an exception was thrown and there is no return value to examine. The only edge case AFAIK would be a write method, since those exceptions may contain a write result within.

Reply:
@Utagai: Your earlier comment in https://github.com/mongodb/mongo-swift-driver/pull/145/files#r235552252 was:

> If self.result is nil, then I expect an empty document instead.

Looking at the current code, if `self.result` is `nil`, `verifyFindAndModifyResult()` returns earlier and doesn't actually assert anything. If the `outcome.result` key is `null` for some reason, I think it's sensible to assert that `distinct()` returns an NSNull value by crafting a `["result": NSNull]` literal. Of course, we know that's not possible so the test would fail.

Reply:
@jmikola I am happy to defer to you on this because I'm not too familiar with the specs or the JSON conventions here, but I'd like to make sure I understand your point:

Are you suggesting that instead of asserting against an empty document `[:]` in the `nil` case, we should just avoid asserting whatsoever, or are you saying we should instead assert against `["result": NSNull()]`?

Reply:
Since [this](https://github.com/mongodb/mongo-swift-driver/pull/145/files#r236788721) clarified what @jmikola wanted and is done, this should also be done.